### PR TITLE
HIVE-26301: Fix ACID tables bootstrap during reverse replication in u…

### PR DESCRIPTION
**Pull request by hmangla98**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 1]
```diff
@@ -1089,9 +1089,8 @@ private boolean needBootstrapAcidTablesDuringIncrementalDump() {
     // of the ACID tables might be included for bootstrap during incremental dump. For old policy, its because the table
     // may not satisfying the old policy but satisfying the new policy. For filter, it may happen that the table
     // is renamed and started satisfying the policy.
-    return ((!work.replScope.includeAllTables())
-            || (previousReplScopeModified())
-            || conf.getBoolVar(HiveConf.ConfVars.REPL_BOOTSTRAP_ACID_TABLES));
+    return !work.replScope.includeAllTables() || previousReplScopeModified() || !tablesForBootstrap.isEmpty()
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:1092]: do we need to update the comment above?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOptimisedBootstrap.java [hunk 2]
```diff
@@ -944,7 +975,8 @@ public void testOverwriteDuringBootstrap() throws Throwable {
 
     // Create some partitioned and non partitioned tables and do a dump & load.
     WarehouseInstance.Tuple tuple = primary.run("use " + primaryDbName)
-        .run("create table t1 (id int)")
+        .run("create table t1 (id int) clustered by(id) into 3 buckets stored as orc " +
+                "tblproperties (\"transactional\"=\"true\")")
```
Comment (pvary) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOptimisedBootstrap.java:979]: do we need to set this? Isn't is the default?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java (hunk 1 above)
Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:1092]: Actually, function name "needBootstrapAcidTablesDuringIncrementalDump" is itself intuitive enough to signify that if tablesForBootstrap is non-empty, we need to bootstrap them.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOptimisedBootstrap.java (hunk 2 above)
Comment (pvary) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOptimisedBootstrap.java:979]: In these tests the default table type is non-ACID?

Comment (hmangla98) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOptimisedBootstrap.java:979]: Default value of hive.strict.managed.tables is false. So, non-Acid managed tables are created if we don't specify the transactional property.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOptimisedBootstrap.java [hunk 3]
```diff
@@ -125,14 +130,8 @@ public void testBuildTableDiffGeneration() throws Throwable {
...
         .run("use " + replicatedDbName)
-        .run("show tables like 't1'")
-        .verifyResult("t1")
-        .run("show tables like 't2'")
-        .verifyResult("t2")
-        .run("show tables like 't1_managed'")
-        .verifyResult("t1_managed")
-        .run("show tables like 't2_managed'")
-        .verifyResult("t2_managed")
+        .run("show tables")
```
Comment (pvary) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOptimisedBootstrap.java:133]: is this change needed?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOptimisedBootstrap.java [hunk 4]
```diff
@@ -962,14 +994,8 @@ public void testOverwriteDuringBootstrap() throws Throwable {
...
-        .run("show tables like 't1'")
-        .verifyResult("t1")
-        .run("show tables like 't2'")
-        .verifyResult("t2")
-        .run("show tables like 't3'")
-        .verifyResult("t3")
-        .run("show tables like 't4'")
-        .verifyResult("t4")
+        .run("show tables")
+        .verifyResults(new String[]{"t1", "t2", "t3", "t4"})
```
Comment (pvary) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOptimisedBootstrap.java:998]: is this change needed?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOptimisedBootstrap.java (hunk 3 above)
Comment (hmangla98) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOptimisedBootstrap.java:133]: We were making unnecessary read calls, even though only one read is required for verification.

**Review by pvary — APPROVED** [review]

**Comment by pvary** [comment]
@hmangla98: Could you please help my review by describing what is in this 2nd commit?

Thanks, Peter

**Comment by hmangla98** [comment]
> @hmangla98: Could you please help my review by describing what is in this 2nd commit?
> 
> Thanks, Peter

Hi @pvary,
This commit is just for test fixes. Actually, some of the test written in TestReplicationOptimisedBootstrap require both primary as well replica cluster to be exclusive (meaning on separate cluster). In this patch, we've modified that functionality to use same cluster. Therefore, we need to move those tests to TestReplicationScenariosExclusiveReplica which mimics the exclusivity of clusters. 

Thanks
Haymant Mangla

**Comment by pvary** [comment]
> Therefore, we need to move those tests to TestReplicationScenariosExclusiveReplica which mimics the exclusivity of clusters.

So basically there is no change in the test code. It is just moved from one test class to another? (Github is not good at handling these moves, it is hard to compare)

**Comment by pvary** [comment]
> > Therefore, we need to move those tests to TestReplicationScenariosExclusiveReplica which mimics the exclusivity of clusters.
> 
> So basically there is no change in the test code. It is just moved from one test class to another? (Github is not good at handling these moves, it is hard to compare)

As discussed offline the tests are only moved

**Review by pvary — APPROVED** [review]
